codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -317,85 +315,96 @@ private Http2Stream requireStream(int streamId) {
-                    int writeablePadding = Math.min(allowedFrameSize - writeableData, padding);
-                    padding -= writeablePadding;
-                    bytesWritten += writeableData + writeablePadding;
-                    ChannelPromise writePromise;
-                    if (size == bytesWritten && !promise.isVoid()) {
-                        // Can use the original promise if it's the last write
-                        writePromise = promise;
+            int maxFrameSize = frameWriter().configuration().frameSizePolicy().maxFrameSize();
+            do {
+                int frameAllowance = Math.min(maxFrameSize, allowedBytes - bytesWritten);
+
+                // Always need a real promise for write as we have to listen to it for failure.
+                ChannelPromise promise = ctx.newPromise();
+                promise.addListener(this);
+                CompositeByteBuf toWrite = ctx.alloc().compositeBuffer(bufsAndPromises.size());
+                while (!bufsAndPromises.isEmpty()) {
+                    final Object entry = bufsAndPromises.remove();
+                    if (entry instanceof ByteBuf) {
+                        ByteBuf buf = (ByteBuf) entry;
+                        if (buf.readableBytes() > frameAllowance) {
+                            ByteBuf slice = buf.readSlice(frameAllowance);
+                            toWrite.addComponent(slice);
+                            toWrite.writerIndex(toWrite.writerIndex() + slice.readableBytes());
+                            // The remaining buffer is kept as a retained slice of the original
+                            bufsAndPromises.addFirst(buf.slice().retain());
+                            frameAllowance = 0;
+                            break;
+                        } else {
+                            toWrite.addComponent(buf);
+                            toWrite.writerIndex(toWrite.writerIndex() + buf.readableBytes());
+                            frameAllowance -= buf.readableBytes();
+                        }
                     } else {
-                        // Create a new promise and listen to it for failure
-                        writePromise = ctx.newPromise();
-                        writePromise.addListener(this);
-                    }
-                    if (toWrite instanceof SlicedByteBuf && data instanceof CompositeByteBuf) {
-                        // If we're writing a subset of a composite buffer then we want to release
-                        // any underlying buffers that have been consumed. CompositeByteBuf only releases
-                        // underlying buffers on write if all of its data has been consumed and its refCnt becomes
-                        // 0.
-                        final CompositeByteBuf toFree = (CompositeByteBuf) data;
-                        writePromise.addListener(new ChannelFutureListener() {
+                        final ChannelPromise nested = (ChannelPromise) entry;
```

Comment:
Agreed. This was getting a little unwieldly and it seemed like the behavior is generic enough that I separated out this functionality into a separate PR. 

Take a look at

https://github.com/netty/netty/pull/3926